### PR TITLE
nixos/nginx: move status page config to `services.nginx`

### PR DESCRIPTION
**Pull request by h7x4**
## Description of changes

The current status page declares itself as the localhost server in raw text, which is problematic if you want to host several endpoints there. I've moved it to `services.nginx.virtualHosts.localhost`, so it will get merged with any other config.

[The nginx nixos tests are currently broken](https://hydra.nixos.org/build/229450287#tabs-summary), and I don't really have the competence to fix them. But I have verified that this change works on my own machine, it behaves as expected, and `nixosTests.prometheus-exporters.nginx` which depends on this specific feature also passes fine.

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by RaitoBezarius — CHANGES_REQUESTED** [review]
Without NixOS test fix *for status page*, I don't think we can proceed with this PR, converting in draft in the meantime.

**Comment by h7x4** [comment]
When you say "test fix for status page", do you mind if I create a test specifically for the status page and ignore the other failures?

**Comment by RaitoBezarius** [comment]
> When you say "test fix for status page", do you mind if I create a test specifically for the status page and ignore the other failures?

Yes, as long as we have a precise test for status page, that's fine by me.

File: nixos/tests/all-tests.nix [hunk 1]
```diff
@@ -535,11 +535,12 @@ in {
   nginx-http3 = handleTest ./nginx-http3.nix {};
   nginx-modsecurity = handleTest ./nginx-modsecurity.nix {};
   nginx-njs = handleTest ./nginx-njs.nix {};
+  nginx-proxyprotocol = handleTest ./nginx-proxyprotocol {};
```
Comment (RaitoBezarius) [inline nixos/tests/all-tests.nix:538]: seems irrelevant to the change, is it?

File: pkgs/servers/http/nginx/generic.nix [hunk 2]
```diff
@@ -178,7 +178,7 @@ stdenv.mkDerivation {
   passthru = {
     inherit modules;
     tests = {
-      inherit (nixosTests) nginx nginx-auth nginx-etag nginx-globalredirect nginx-http3 nginx-pubhtml nginx-sandbox nginx-sso nginx-proxyprotocol;
+      inherit (nixosTests) nginx nginx-auth nginx-etag nginx-globalredirect nginx-http3 nginx-proxyprotocol nginx-pubhtml nginx-sandbox nginx-sso nginx-status-page;
```
Comment (RaitoBezarius) [inline pkgs/servers/http/nginx/generic.nix:181]: irrelevant to the change too

File: nixos/modules/services/web-servers/nginx/default.nix [hunk 3]
```diff
@@ -1175,6 +1158,21 @@ in
     services.nginx.additionalModules = optional cfg.recommendedBrotliSettings pkgs.nginxModules.brotli
       ++ lib.optional cfg.recommendedZstdSettings pkgs.nginxModules.zstd;
 
+    services.nginx.virtualHosts.localhost = mkIf cfg.statusPage {
+      listenAddresses = [
```
Comment (RaitoBezarius) [inline nixos/modules/services/web-servers/nginx/default.nix:1162]:
```suggestion
      listenAddresses = lib.mkDefault [
```

probably?

**Review by RaitoBezarius — APPROVED** [review]
Sounds good, please move the irrelevant change in their own commit, otherwise this will pollute the history.

**Comment by h7x4** [comment]
I've moved it to a third commit. Ready for merge

**Review by RaitoBezarius — APPROVED** [review]

**Comment by RaitoBezarius** [comment]
@ofborg test nginx-status-page
